File: js/packages/app/src/__tests__/features/object-references/external/polywrap/object-references-example/src/polywrap/wrapped/TestObjectGetter/invoke.ts
import { bufferToU32, BaseTypeSerialization, IExternalWrapInstance } from "@polywrap/reim-wrap-js";
import { WrapManifest } from '../../WrapManifest';
import { TestObjectGetterWrapped } from "..";
import { TestObjectGetter } from "../../external/classes";

import { TestExternalClassWrapped } from "..";

import { TestExternalClass } from "../../..";



export function invoke(buffer: Uint8Array, wrapInstance: IExternalWrapInstance): Promise<Uint8Array> {
  const funcId = bufferToU32(buffer);
  const dataBuffer = buffer.slice(4);

  switch (funcId) {
    case WrapManifest.Internal.Classes.TestObjectGetterMethod.Create:
        return invokeCreateWrapped(dataBuffer, wrapInstance);
    case WrapManifest.Internal.Classes.TestObjectGetterMethod.TestInstanceReceiveReference:
        return invokeTestInstanceReceiveReferenceWrapped(dataBuffer, wrapInstance);
    case WrapManifest.Internal.Classes.TestObjectGetterMethod.TestStaticReceiveReference:
        return invokeTestStaticReceiveReferenceWrapped(dataBuffer, wrapInstance);
    default:
      throw new Error(`Unknown method: ${funcId.toString()} on class `);
  }
}

const invokeCreateWrapped = async (buffer: Uint8Array, wrapInstance: IExternalWrapInstance): Promise<Uint8Array> => {
  const args = CreateArgsWrapped.deserialize(buffer, wrapInstance);

  const result = await TestObjectGetter.create(
    args.arg,
  );
  
  return TestObjectGetterWrapped.serialize(result);
  };

class CreateArgs {
  constructor(
    public arg: string,
  ) {
  }
}

class CreateArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: Uint8Array, wrapInstance: IExternalWrapInstance): CreateArgs {
    const args = JSON.parse(new TextDecoder().decode(buffer));
  
    return new CreateArgs(
            
      args.arg,
      
    );
  }  
}
const invokeTestInstanceReceiveReferenceWrapped = async (buffer: Uint8Array, wrapInstance: IExternalWrapInstance): Promise<Uint8Array> => {
  
  const referencePtr = bufferToU32(buffer);
  const dataBuffer = buffer.slice(4);

  const args = TestInstanceReceiveReferenceArgsWrapped.deserialize(dataBuffer, wrapInstance);

  const object = TestObjectGetterWrapped.dereference(referencePtr);

  const result = await object.testInstanceReceiveReference(
    args.arg,
  );

  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestInstanceReceiveReferenceArgs {
  constructor(
    public arg: TestExternalClass,
  ) {
  }
}

class TestInstanceReceiveReferenceArgsWrapped {
  constructor(
    public arg: TestExternalClassWrapped,
  ) {
  }

  static deserialize(buffer: Uint8Array, wrapInstance: IExternalWrapInstance): TestInstanceReceiveReferenceArgs {
    const args = JSON.parse(new TextDecoder().decode(buffer));
  
    return new TestInstanceReceiveReferenceArgs(
      
      TestExternalClassWrapped.mapFromSerializable(args.arg, wrapInstance),
            
    );
  }  
}
const invokeTestStaticReceiveReferenceWrapped = async (buffer: Uint8Array, wrapInstance: IExternalWrapInstance): Promise<Uint8Array> => {
  const args = TestStaticReceiveReferenceArgsWrapped.deserialize(buffer, wrapInstance);

  const result = await TestObjectGetter.testStaticReceiveReference(
    args.arg,
  );
  
  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestStaticReceiveReferenceArgs {
  constructor(
    public arg: TestExternalClass,
  ) {
  }
}

class TestStaticReceiveReferenceArgsWrapped {
  constructor(
    public arg: TestExternalClassWrapped,
  ) {
  }

  static deserialize(buffer: Uint8Array, wrapInstance: IExternalWrapInstance): TestStaticReceiveReferenceArgs {
    const args = JSON.parse(new TextDecoder().decode(buffer));
  
    return new TestStaticReceiveReferenceArgs(
      
      TestExternalClassWrapped.mapFromSerializable(args.arg, wrapInstance),
            
    );
  }  
}

File: rust/app/src/polywrap/external_references/src/polywrap/wrapped/TestObjectGetter/invoke.rs
use reim_wrap::{ ExternalWrapModule };
use module::{ WrapManifest };
use internal::classes::{ TestObjectGetterWrapped };
use external::classes::{ TestObjectGetter };

use internal::classes::{ TestExternalClassWrapped };

use crate::{ TestExternalClass };



pub async fn invoke(buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
  let func_id = u32::from_be_bytes(buffer[0..4].try_into().expect("Method ID must be 4 bytes"));
  let data_buffer = &buffer[4..];

  match func_id {
    wrap_manifest::internal::classes::TestObjectGetterMethod::Create =>
        invokeCreateWrapped(data_buffer, external_module),
    wrap_manifest::internal::classes::TestObjectGetterMethod::TestInstanceReceiveReference =>
        invokeTestInstanceReceiveReferenceWrapped(data_buffer, external_module),
    wrap_manifest::internal::classes::TestObjectGetterMethod::TestStaticReceiveReference =>
        invokeTestStaticReceiveReferenceWrapped(data_buffer, external_module),
    _ => panic!(format!("Unknown method: {} on class {}", func_id.to_string(), model.class_name))
  }
}

async fn invokeCreateWrapped(buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
    let args = CreateArgsWrapped.deserialize(buffer, external_module);

    let result = TestObjectGetter.create(
        args.arg,
    ).await;
    
    TestObjectGetterWrapped.serialize(result)
    };

struct CreateArgs {
    pub arg: string,
}

impl CreateArgs {
    pub fn new(
        arg: string,
    ) -> Self {
        Self {
            arg,
        }
    }
}

struct CreateArgsWrapped {
    pub arg: string,
}

impl CreateArgsWrapped {
    pub fn new(
        arg: string,
    ) {
        Self {
        arg,
        }
    }

    pub fn deserialize(buffer: Uint8Array, external_module: dyn ExternalWrapModule) -> CreateArgs {
        let args = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
    
        CreateArgs.new(
                
        args.arg,
        
        )
    }  
}
async fn invokeTestInstanceReceiveReferenceWrapped(buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
    
    let reference_ptr = u32::from_be_bytes(buffer[0..4].try_into().expect("Reference ptr must be 4 bytes"));
    let data_buffer = &buffer[4..];

    let args = TestInstanceReceiveReferenceArgsWrapped.deserialize(data_buffer, external_module);

    let object = TestObjectGetterWrapped.dereference(reference_ptr);

    let result = object.testInstanceReceiveReference(
        args.arg,
    ).await;

    
    BaseTypeSerialization.serialize<string>(result)
};

struct TestInstanceReceiveReferenceArgs {
    pub arg: TestExternalClass,
}

impl TestInstanceReceiveReferenceArgs {
    pub fn new(
        arg: TestExternalClass,
    ) -> Self {
        Self {
            arg,
        }
    }
}

struct TestInstanceReceiveReferenceArgsWrapped {
    pub arg: TestExternalClassWrapped,
}

impl TestInstanceReceiveReferenceArgsWrapped {
    pub fn new(
        arg: TestExternalClassWrapped,
    ) {
        Self {
        arg,
        }
    }

    pub fn deserialize(buffer: Uint8Array, external_module: dyn ExternalWrapModule) -> TestInstanceReceiveReferenceArgs {
        let args = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
    
        TestInstanceReceiveReferenceArgs.new(
        
        TestExternalClassWrapped.map_from_serializable(args.arg, external_module),
                
        )
    }  
}
async fn invokeTestStaticReceiveReferenceWrapped(buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
    let args = TestStaticReceiveReferenceArgsWrapped.deserialize(buffer, external_module);

    let result = TestObjectGetter.testStaticReceiveReference(
        args.arg,
    ).await;
    
    
    BaseTypeSerialization.serialize<string>(result)
};

struct TestStaticReceiveReferenceArgs {
    pub arg: TestExternalClass,
}

impl TestStaticReceiveReferenceArgs {
    pub fn new(
        arg: TestExternalClass,
    ) -> Self {
        Self {
            arg,
        }
    }
}

struct TestStaticReceiveReferenceArgsWrapped {
    pub arg: TestExternalClassWrapped,
}

impl TestStaticReceiveReferenceArgsWrapped {
    pub fn new(
        arg: TestExternalClassWrapped,
    ) {
        Self {
        arg,
        }
    }

    pub fn deserialize(buffer: Uint8Array, external_module: dyn ExternalWrapModule) -> TestStaticReceiveReferenceArgs {
        let args = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
    
        TestStaticReceiveReferenceArgs.new(
        
        TestExternalClassWrapped.map_from_serializable(args.arg, external_module),
                
        )
    }  
}
